quesma/quesma/search.go

```diff
@@ -213,129 +333,101 @@ func (q *QueryRunner) handleSearchCommon(ctx context.Context, indexPattern strin
-					if err != nil {
-						doneCh <- AsyncSearchWithError{translatedQueryBody: translatedQueryBody, err: err}
-					}
-				}
+			// results are passed via channel
+			q.executePlan(ctx, alternativePlan, queryTranslator, table, body, optAsync, executionChan)
 
-				searchResponse := queryTranslator.MakeSearchResponse(plan.Queries, results)
+		}()
 
-				doneCh <- AsyncSearchWithError{response: searchResponse, translatedQueryBody: translatedQueryBody, err: err}
-			}()
+		go func(executionChan chan executionPlanResult) {
+			defer recovery.LogPanic()
+			var alternative executionPlanResult
+			var main executionPlanResult
 
-		} else {
-			queries := plan.Queries
-			queriesBody := make([]types.TranslatedSQLQuery, len(queries))
-			queriesBodyConcat := ""
-			for i, query := range queries {
-				queriesBody[i].Query = []byte(query.SelectCommand.String())
-				queriesBodyConcat += query.SelectCommand.String() + "\n"
-			}
-			responseBody = []byte(fmt.Sprintf("Invalid Queries: %v, err: %v", queriesBody, err))
-			logger.ErrorWithCtxAndReason(ctx, "Quesma generated invalid SQL query").Msg(queriesBodyConcat)
-			bodyAsBytes, _ := body.Bytes()
-			pushSecondaryInfo(q.quesmaManagementConsole, id, "", path, bodyAsBytes, queriesBody, responseBody, startTime)
-			return responseBody, errors.New(string(responseBody))
-		}
-
-		if optAsync == nil {
-			bodyAsBytes, _ := body.Bytes()
-			response := <-doneCh
-			if response.err != nil {
-				err = response.err
-				if len(plan.Queries) > 0 {
-					logger.ErrorWithCtx(ctx).Msgf("error making response: %v, queries[0]: %+v", err, plan.Queries[0])
-				} else {
-					logger.ErrorWithCtx(ctx).Msgf("error making response: %v, queries empty", err)
+			// we have only two plans to execute
+			for range 2 {
+				r := <-executionChan
+				logger.InfoWithCtx(ctx).Msgf("received results  %s", r.plan.Name)
+				if r.plan.Name == model.AlternativeExecutionPlan {
+					alternative = r
+				} else if r.plan.Name == model.MainExecutionPlan {
+					main = r
 				}
-			} else {
-				responseBody, err = response.response.Marshal()
 			}
-			pushSecondaryInfo(q.quesmaManagementConsole, id, "", path, bodyAsBytes, response.translatedQueryBody, responseBody, startTime)
-			return responseBody, err
-		} else {
-			select {
-			case <-time.After(time.Duration(optAsync.waitForResultsMs) * time.Millisecond):
-				go func() { // Async search takes longer. Return partial results and wait for
-					recovery.LogPanicWithCtx(ctx)
-					res := <-doneCh
-					q.storeAsyncSearch(q.quesmaManagementConsole, id, optAsync.asyncId, optAsync.startTime, path, body, res, true)
-				}()
-				return q.handlePartialAsyncSearch(ctx, optAsync.asyncId)
-			case res := <-doneCh:
-				responseBody, err = q.storeAsyncSearch(q.quesmaManagementConsole, id, optAsync.asyncId, optAsync.startTime, path, body, res,
-					optAsync.keepOnCompletion)
-
-				return responseBody, err
+
+			if string(alternative.responseBody) != string(main.responseBody) {
```

Comment:
My assumption is that having two plans is only intermediate state, correct?. So, at some point we should have only one way of building and executing queries

Reply:
Generally speaking, it's a framework for A/B testing on production. Right now we're testing correctness only. But we may test performance, and resource consumption as well.

Reply:
We may want to signal it to quesma console log.

Reply:
Maybe we should use `util.JsonDifference` to check if there are differences.

Also I hope that if you print error it would showed up the console.